TS-4834 Expose bad disk and disk access failures

gtenev opened the pull request:
For monitoring purposes expose stats about:
- disk access failure count
- marked bad disk count

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/755/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/651/ for details.

jpeach commented:
My comments from [TS-4834](https://issues.apache.org/jira/browse/TS-4834) ..

>I think the nomenclature should mirror the existing metrics which typically don't include "count" in the names. Especially since the count of bad disks (gauge) does not have the same semantics as the count of disk errors (counter).

Looking at the code I think the bad disk metric is a counter not a gauge.

> Is "disk" the right nomenclature, or would "span" be more accurate?
How feasible is it to have separate metrics for read and write errors? Does it make sense to also have metrics for disk read and write operations, or are system tools sufficient for that?
If you have a count of bad disks, does it make sense to have other disk counts? ie. good disks, offline disks, etc.

SolidWallOfCode commented on iocore/cache/Cache.cc:
```diff
@@ -2070,6 +2070,16 @@ AIO_Callback_handler::handle_disk_failure(int /* event ATS_UNUSED */, void *data
     CacheDisk *d = gdisks[disk_no];
 
+    /* find the Vol instance needed for updating per-volume stats by CACHE_INCREMENT_DYN_STAT */
+    Vol *vol;
```
If there are multiple volumes for the same physical disk / file descriptor, does this just find the first one?

jpeach commented:
@gtenev See my comments here and in the bug. I think that these new metrics should follow the existing conventions of the cache system.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/829/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/725/ for details.

gtenev commented:
@jpeach and @SolidWallOfCode, really appreciate your feedback! Here is the new patch.

Removed `_count`. The reason I added was that internally the "bad disk metric" and "disk errors" both had counter semantics and there are already metrics containing `_count`.

Renamed `disk` to `span`, this seems more accurate indeed.

I like the idea of exposing "online", "offline", etc. The new patch exposes the following metrics as gauges: 
```
proxy.process.cache.span.failing
proxy.process.cache.span.offline
proxy.process.cache.span.online
```
 A span "moves" from "online" bucket (`errors==0`) to "failing" (`errors > 0 && errors < proxy.config.cache.max_disk_errors`) to "offline" (`errors >= proxy.config.cache.max_disk_errors`).

Please note that "failing" + "offline" + "online" = total number of spans.

It was possible to split the read and write metrics so removed `proxy.process.cache.disk.errors` in favor of
```
proxy.process.cache.span.errors.read
proxy.process.cache.span.errors.write
```

Removed the "per-volume" stats. @SolidWallOfCode, incrementing the metrics for all volumes on each failure does not make sense either. I would need to add more code to be able to increment the metrics for the right volume and I am not sure it is worth the effort (and maintenance).

I think the idea of this change in general is to give the operator a signal that the disks need to be inspected and then the operator would diagnose them with more sophisticated lower level disk utilities.

Noticed that the metrics would not get updated properly (would get somehow inconsistent) if there were failures during the cache initialization (before "cache enabled") and tried to fix wherever  noticed and able to test / reproduce.

zwoop commented:
@gtenev This is ready to land now, right ?

gtenev commented:
@zwoop, I have not heard any objections for a while so unless @jpeach and @SolidWallOfCode have any concerns with the latest patch I think we can land it.

SolidWallOfCode commented on iocore/cache/CacheDisk.cc:
```diff
@@ -23,6 +23,35 @@
+  this->num_errors++;
+
+  int opcode   = io->aiocb.aio_lio_opcode;
+  char *opname = (char *)"unknown";
```
Why not declare this as `const char*` instead of casting?

SolidWallOfCode commented on iocore/cache/Cache.cc:
```diff
@@ -2015,6 +2022,16 @@ CacheProcessor::mark_storage_offline(CacheDisk *d ///< Target disk
   RecIncrGlobalRawStat(cache_rsb, cache_direntries_used_stat, -used_dir_delete);
 
+  /* Update the span metrics */
+  if (!admin) {
```
Maybe
```
RecIncrGlobalRawStats(cache_rsb, admin ? cache_span_offline_stat, cache_span_failing_state, -1); 
```

SolidWallOfCode commented:
Are the casts on the state indices required?

SolidWallOfCode commented:
@gtenev Just do the tweaks and we'll commit this.

gtenev commented:
@jpeach added docs, @SolidWallOfCode made the requested change.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1124/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1019/ for details.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1126/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1020/ for details.

atsci commented:
FreeBSD build *successful*! See https://ci.trafficserver.apache.org/job/Github-FreeBSD/1127/ for details.

atsci commented:
Linux build *successful*! See https://ci.trafficserver.apache.org/job/Github-Linux/1021/ for details.

gtenev commented:
@zwoop, it is ready to land now.
